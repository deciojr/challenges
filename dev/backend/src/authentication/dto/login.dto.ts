import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString } from 'class-validator';

import { DTO } from '../../shared/model/dto.model';

export class LoginDTO implements Readonly<DTO<LoginDTO>> {
  @ApiProperty({
    required: true,
    description:
      'Authentication header containing the user email and password using the format ```email:password``` encoded with base64',
  })
  @IsString()
  @IsNotEmpty()
  authentication: string;
}
