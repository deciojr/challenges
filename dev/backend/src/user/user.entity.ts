import { Audit } from '../shared/model/audit.model';
import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';

import * as bcrypt from 'bcrypt';

@Entity({ name: 'users' })
export class User extends Audit {
  @PrimaryGeneratedColumn('uuid')
  uuid?: string;

  @Column()
  firstName: string;

  @Column()
  lastName: string;

  @Column({ unique: true })
  email: string;

  @Column()
  password: string;

  @Column()
  @Index()
  secretQuestion: string;

  @Column()
  @Index()
  secretAnswer: string;

  async hashPassword?(password: string) {
    const rounds = 10;
    return await bcrypt.hash(password, rounds);
  }

  async comparePasswords?(password, encryptedPassword) {
    return await bcrypt.compare(password, encryptedPassword);
  }
}
