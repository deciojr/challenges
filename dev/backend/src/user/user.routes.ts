import { Routes } from 'src/shared/interface/routes.inteface';

export const userRoutes: Routes = {
  prefix: 'user',
  swaggerTag: 'User',
  paths: {
    save: '',
  },
};
