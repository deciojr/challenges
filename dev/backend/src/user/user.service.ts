import { HttpStatus, Inject, Injectable } from '@nestjs/common';

import { User } from './user.entity';
import { UserRepository } from './user.repository';

@Injectable()
export class UserService {
  constructor(
    @Inject(UserRepository) readonly userRepository: UserRepository,
  ) {}

  async save(user: User): Promise<HttpStatus> {
    try {
      const isEmailRegistered = (await this.countByEmail(user.email)) > 0;
      if (isEmailRegistered) {
        return HttpStatus.CONFLICT;
      }
      await this.userRepository.save(user);
      return HttpStatus.CREATED;
    } catch (e) {
      return HttpStatus.BAD_REQUEST;
    }
  }

  async countByEmail(email: string): Promise<number> {
    return await this.userRepository.count({
      email,
    });
  }
}
