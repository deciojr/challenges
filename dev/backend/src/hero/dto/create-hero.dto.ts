import { ApiProperty } from '@nestjs/swagger/';
import { IsEnum, IsNotEmpty, IsString } from 'class-validator';

import { DTO } from '../../shared/model/dto.model';
import { HeroClass, Hero } from '../hero.entity';

export class CreateHeroDTO implements Readonly<DTO<CreateHeroDTO>> {
  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  name: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  badge: string;

  @ApiProperty({ required: true, enum: HeroClass })
  @IsEnum(HeroClass)
  @IsNotEmpty()
  heroClass: HeroClass;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  lat: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  lng: string;

  static async toHero({
    name,
    badge,
    heroClass,
    lat,
    lng,
  }: CreateHeroDTO): Promise<Hero> {
    const hero = new Hero();

    hero.name = name;
    hero.badge = badge;
    hero.heroClass = heroClass;
    hero.lat = lat;
    hero.lng = lng;

    return hero;
  }
}
