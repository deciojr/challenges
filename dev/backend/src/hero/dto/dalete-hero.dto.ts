import { ApiProperty } from '@nestjs/swagger/';
import { IsEmail, IsNotEmpty, IsString } from 'class-validator';

import { DTO } from '../../shared/model/dto.model';
import { HeroClass, Hero } from '../hero.entity';

export class DeleteHeroDTO implements Readonly<DTO<DeleteHeroDTO>> {
  @ApiProperty({ required: true })
  id: string;
}
