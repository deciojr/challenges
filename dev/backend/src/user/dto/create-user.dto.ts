import { ApiProperty } from '@nestjs/swagger/';
import { IsEmail, IsNotEmpty, IsString } from 'class-validator';
import { DTO } from '../../shared/model/dto.model';

import { User } from '../user.entity';

export class CreateUserDTO implements Readonly<DTO<CreateUserDTO>> {
  static examples: {
    valid: CreateUserDTO;
    invalid: CreateUserDTO;
  } = {
    valid: {
      email: 'email@email.com',
      password: 'password',
      firstName: 'firstname',
      lastName: 'lastname',
      secretQuestion: 'question',
      secretAnswer: 'answer',
    },
    invalid: {
      email: undefined,
      password: '123',
      firstName: undefined,
      lastName: undefined,
      secretQuestion: undefined,
      secretAnswer: undefined,
    },
  };

  @ApiProperty({ required: true })
  @IsEmail()
  @IsString()
  @IsNotEmpty()
  email: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  firstName: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  lastName: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  password: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  secretQuestion: string;

  @ApiProperty({ required: true })
  @IsString()
  @IsNotEmpty()
  secretAnswer: string;

  static async toUser({
    firstName,
    lastName,
    email,
    password,
    secretQuestion,
    secretAnswer,
  }: CreateUserDTO): Promise<User> {
    const user = new User();

    user.firstName = firstName;
    user.lastName = lastName;
    user.password = await user.hashPassword(password);
    user.email = email;
    user.secretQuestion = secretQuestion;
    user.secretAnswer = secretAnswer;

    return user;
  }
}
