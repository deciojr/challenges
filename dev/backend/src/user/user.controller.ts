import {
  Body,
  Controller,
  HttpStatus,
  Inject,
  Post,
  Res,
} from '@nestjs/common';

import { Response } from 'express';

import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';

import { userRoutes } from './user.routes';
import { CreateUserDTO } from './dto/create-user.dto';
import { UserService } from './user.service';

@Controller(userRoutes.prefix)
@ApiTags(userRoutes.swaggerTag)
export class UserController {
  constructor(@Inject(UserService) readonly userService: UserService) {}

  @ApiOperation({
    description: 'Create a new User',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'User created',
  })
  @ApiResponse({
    status: HttpStatus.BAD_REQUEST,
    description: 'Could not create the user',
  })
  @ApiResponse({
    status: HttpStatus.CONFLICT,
    description: 'Email already registered',
  })
  @Post(userRoutes.paths.save as string)
  async save(@Body() createUserDTO: CreateUserDTO, @Res() response: Response) {
    const user = await CreateUserDTO.toUser(createUserDTO);

    const statusCode = await this.userService.save(user);

    return response.send({
      statusCode,
    });
  }
}
