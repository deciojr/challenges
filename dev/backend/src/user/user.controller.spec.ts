import { HttpStatus } from '@nestjs/common';
import { Connection } from 'typeorm';
import { createInMemoryDatabaseConnection } from '../shared/test/create-in-memory-database-connection';
import { CreateUserDTO } from './dto/create-user.dto';

import { UserController } from './user.controller';
import { User } from './user.entity';
import { UserRepository } from './user.repository';
import { UserService } from './user.service';

describe('User Controller', () => {
  let controller: UserController;
  let service: UserService;
  let repository: UserRepository;
  let database: Connection;

  const response: any = {
    send: (body?: any) => body,
  };

  beforeAll(async () => {
    database = await createInMemoryDatabaseConnection([User], 'userController');
    repository = database.getRepository(User);
    service = new UserService(repository);
    controller = new UserController(service);
  });

  describe('Save a User', () => {
    it('should return status code 200 on success', async () => {
      expect(
        await controller.save(CreateUserDTO.examples.valid, response),
      ).toMatchSnapshot({
        statusCode: HttpStatus.CREATED,
      });
    });
    it('should return status code 409 on malformed request (invalid DTO)', async () => {
      expect(
        await controller.save(CreateUserDTO.examples.invalid, response),
      ).toMatchSnapshot({
        statusCode: HttpStatus.BAD_REQUEST,
      });
    });

    it('should return status code 401 on email conflict ', async () => {
      await controller.save(CreateUserDTO.examples.valid, response);

      expect(
        await controller.save(CreateUserDTO.examples.valid, response),
      ).toMatchSnapshot({
        statusCode: HttpStatus.CONFLICT,
      });
    });
  });
});
