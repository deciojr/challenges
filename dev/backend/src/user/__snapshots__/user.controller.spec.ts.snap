// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Controller Save a User should return status code 200 on success 1`] = `
Object {
  "statusCode": 201,
}
`;

exports[`User Controller Save a User should return status code 401 on email conflict  1`] = `
Object {
  "statusCode": 409,
}
`;

exports[`User Controller Save a User should return status code 409 on malformed request (invalid DTO) 1`] = `
Object {
  "statusCode": 400,
}
`;
